### PR TITLE
Add substudy and external ID relationships to the HealthData table.

### PR DESCRIPTION
**Pull request by alxdarksage**

### CODE REVIEW

File: app/org/sagebionetworks/bridge/dynamodb/DynamoHealthDataRecord.java [hunk 1]
```diff
@@ -295,6 +297,19 @@ public void setUserDataGroups(Set<String> userDataGroups) {
         // DDB doesn't support empty sets, use null reference for empty set. This is also enforced by the builder.
         this.userDataGroups = (userDataGroups != null && !userDataGroups.isEmpty()) ? userDataGroups : null;
     }
+    
+    /** {@inheritDoc} */
+    @DynamoDBTypeConverted(converter=StringValueStringKeyMapMarshaller.class)
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/dynamodb/DynamoHealthDataRecord.java]: Dynamo handles Maps just fine. Is there any particular reason we need to marshal to/from JSON?

File: app/org/sagebionetworks/bridge/dynamodb/StringValueStringKeyMapMarshaller.java [hunk 2]
```diff
@@ -0,0 +1,14 @@
+package org.sagebionetworks.bridge.dynamodb;
+
+import java.util.Map;
+
+import com.fasterxml.jackson.core.type.TypeReference;
+
+public class StringValueStringKeyMapMarshaller extends StringKeyMapMarshaller<String> {
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/dynamodb/StringValueStringKeyMapMarshaller.java]: Dynamo handles Maps just fine. Is there any particular reason we need to marshal to/from JSON?

**Review by DwayneJengSage — CHANGES_REQUESTED** [review]
Looks good. One comment about whether we need a String map marshaller.

File: app/org/sagebionetworks/bridge/dynamodb/DynamoHealthDataRecord.java (hunk 1 above)
Comment (alxdarksage) [reply app/org/sagebionetworks/bridge/dynamodb/DynamoHealthDataRecord.java]: @DynamoDBTypeConvertedJson works perfectly fine, deleted this custom type converter in favor of that. Integration tests pass.

File: app/org/sagebionetworks/bridge/dynamodb/StringValueStringKeyMapMarshaller.java (hunk 2 above)
Comment (alxdarksage) [reply app/org/sagebionetworks/bridge/dynamodb/StringValueStringKeyMapMarshaller.java]: Switched to @DynamoDBTypeConvertedJson.

File: app/org/sagebionetworks/bridge/dynamodb/DynamoHealthDataRecord.java [hunk 3]
```diff
@@ -295,6 +298,19 @@ public void setUserDataGroups(Set<String> userDataGroups) {
         // DDB doesn't support empty sets, use null reference for empty set. This is also enforced by the builder.
         this.userDataGroups = (userDataGroups != null && !userDataGroups.isEmpty()) ? userDataGroups : null;
     }
+    
+    /** {@inheritDoc} */
+    @DynamoDBTypeConvertedJson
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/dynamodb/DynamoHealthDataRecord.java]: I think there's still a misunderstanding here. Why do we need JSON at all? DynamoDB handles String maps just fine.

**Review by DwayneJengSage — CHANGES_REQUESTED** [review]

File: app/org/sagebionetworks/bridge/dynamodb/DynamoHealthDataRecord.java (hunk 3 above)
Comment (alxdarksage) [reply app/org/sagebionetworks/bridge/dynamodb/DynamoHealthDataRecord.java]: Not according to the integration tests. Without this annotation the map is always empty.

Comment (DwayneJengSage) [reply app/org/sagebionetworks/bridge/dynamodb/DynamoHealthDataRecord.java]: It seems to work just fine for DynamoStudy.getInstallLinks

Comment (alxdarksage) [reply app/org/sagebionetworks/bridge/dynamodb/DynamoHealthDataRecord.java]: Verified (some part of BridgePF, BridgeJavaSDK and/or BridgeIntegrationTests must not have been in the right branch or updated when I tested this the first time... it persists fine).

**Comment by alxdarksage** [comment]
This was updated to remove the marshaller.

**Review by DwayneJengSage — APPROVED** [review]

**Pull request merged**